File: backend/src/main/java/fr/esgi/api/verydarkkitchensmartphone/service/ReservationService.java
package fr.esgi.api.verydarkkitchensmartphone.service;

import fr.esgi.api.verydarkkitchensmartphone.dto.AvailabilityRequest;
import fr.esgi.api.verydarkkitchensmartphone.dto.AvailabilityResponse;
import fr.esgi.api.verydarkkitchensmartphone.dto.PlatDTO;
import fr.esgi.api.verydarkkitchensmartphone.dto.ReservationRequest;
import fr.esgi.api.verydarkkitchensmartphone.dto.ReservationResponse;
import fr.esgi.api.verydarkkitchensmartphone.dto.UpdateReservationRequest;
import fr.esgi.api.verydarkkitchensmartphone.models.Plat;
import fr.esgi.api.verydarkkitchensmartphone.models.Reservation;
import fr.esgi.api.verydarkkitchensmartphone.models.StatutReservation;
import fr.esgi.api.verydarkkitchensmartphone.models.User;
import fr.esgi.api.verydarkkitchensmartphone.repository.PlatRepository;
import fr.esgi.api.verydarkkitchensmartphone.repository.ReservationRepository;
import fr.esgi.api.verydarkkitchensmartphone.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Service gérant les opérations métier liées aux réservations.
 * Permet de créer, consulter, modifier et annuler des réservations de restaurant.
 */
@Service
@RequiredArgsConstructor
@Transactional
public class ReservationService {

    private final ReservationRepository reservationRepository;
    private final PlatRepository platRepository;
    private final UserRepository userRepository;
    private final AvailabilityService availabilityService;

    /**
     * Crée une nouvelle réservation dans le système.
     * Initialise le statut à EN_ATTENTE et enregistre la date de création.
     * Vérifie la disponibilité avant de créer la réservation.
     *
     * @param request Données de la réservation (client, date, plats, etc.)
     * @return Réponse contenant les détails de la réservation créée
     * @throws RuntimeException si pas assez de places disponibles
     */
    public ReservationResponse createReservation(ReservationRequest request) {
        // Vérification de la disponibilité
        AvailabilityRequest availabilityRequest = new AvailabilityRequest(
                request.getDateReservation(),
                request.getNombrePersonnes()
        );
        AvailabilityResponse availability = availabilityService.checkAvailability(availabilityRequest);

        if (!availability.getAvailable()) {
            throw new RuntimeException(
                String.format("Pas assez de places disponibles. Places disponibles: %d, demandées: %d",
                    availability.getAvailableSeats(),
                    request.getNombrePersonnes())
            );
        }

        // Récupération des plats sélectionnés
        List<Plat> plats = platRepository.findAllById(request.getPlatIds());

        // Construction de l'entité réservation
        Reservation reservation = Reservation.builder()
                .user(userRepository.findById(request.getIdClient()).orElse(null))
                .email(request.getEmail())
                .telephone(request.getTelephone())
                .dateReservation(request.getDateReservation())
                .nombrePersonnes(request.getNombrePersonnes())
                .plats(plats)
                .commentaire(request.getCommentaire())
                .statut(StatutReservation.EN_ATTENTE)
                .dateCreation(LocalDateTime.now())
                .build();

        Reservation saved = reservationRepository.save(reservation);
        return convertToResponse(saved);
    }

    /**
     * Récupère une réservation spécifique par son identifiant.
     *
     * @param id Identifiant de la réservation
     * @return Détails de la réservation
     * @throws RuntimeException si la réservation n'est pas trouvée
     */
    public ReservationResponse getReservationById(Long id) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Réservation non trouvée"));
        return convertToResponse(reservation);
    }

    /**
     * Récupère toutes les réservations enregistrées dans le système.
     *
     * @return Liste de toutes les réservations
     */
    public List<ReservationResponse> getAllReservations() {
        return reservationRepository.findAll().stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Récupère toutes les réservations associées à une adresse email.
     * Utile pour les clients non authentifiés ou les recherches par email.
     *
     * @param email Adresse email du client
     * @return Liste des réservations liées à cet email
     */
    public List<ReservationResponse> getReservationsByEmail(String email) {
        return reservationRepository.findByEmail(email).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Récupère toutes les réservations d'un utilisateur authentifié.
     *
     * @param userId Identifiant de l'utilisateur
     * @return Liste des réservations de l'utilisateur
     * @throws java.util.NoSuchElementException si l'utilisateur n'existe pas
     */
    public List<ReservationResponse> getReservationByUserId(Long userId) {
        User user = userRepository.findById(userId).orElseThrow();
        return reservationRepository.findByUser(user).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Met à jour le statut d'une réservation existante.
     * Permet de passer une réservation de EN_ATTENTE à CONFIRMEE, ANNULEE, etc.
     *
     * @param id Identifiant de la réservation
     * @param statut Nouveau statut à appliquer (doit correspondre à un StatutReservation)
     * @return Réservation mise à jour
     * @throws RuntimeException si la réservation n'est pas trouvée
     * @throws IllegalArgumentException si le statut fourni n'est pas valide
     */
    public ReservationResponse updateStatut(Long id, String statut) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Réservation non trouvée"));
        reservation.setStatut(StatutReservation.valueOf(statut));
        return convertToResponse(reservationRepository.save(reservation));
    }

    /**
     * Annule une réservation en changeant son statut à ANNULEE.
     * La réservation n'est pas supprimée de la base de données pour conserver l'historique.
     *
     * @param id Identifiant de la réservation à annuler
     * @throws RuntimeException si la réservation n'est pas trouvée
     */
    public void cancelReservation(Long id) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Réservation non trouvée"));
        reservation.setStatut(StatutReservation.ANNULEE);
        reservationRepository.save(reservation);
    }

    /**
     * Met à jour une réservation existante avec de nouvelles informations.
     * Vérifie la disponibilité si la date ou le nombre de personnes change.
     *
     * @param id Identifiant de la réservation
     * @param request Nouvelles données de réservation
     * @return Réservation mise à jour
     * @throws RuntimeException si la réservation n'est pas trouvée ou si pas assez de places
     */
    public ReservationResponse updateReservation(Long id, UpdateReservationRequest request) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Réservation non trouvée"));

        // Vérifier si la date ou le nombre de personnes change
        boolean needsAvailabilityCheck = false;
        LocalDateTime newDate = request.getDateReservation() != null ?
            request.getDateReservation() : reservation.getDateReservation();
        Integer newNombrePersonnes = request.getNombrePersonnes() != null ?
            request.getNombrePersonnes() : reservation.getNombrePersonnes();

        if (!newDate.equals(reservation.getDateReservation()) ||
            !newNombrePersonnes.equals(reservation.getNombrePersonnes())) {
            needsAvailabilityCheck = true;
        }

        // Vérifier la disponibilité si nécessaire
        if (needsAvailabilityCheck) {
            AvailabilityRequest availabilityRequest = new AvailabilityRequest(
                newDate, newNombrePersonnes
            );
            AvailabilityResponse availability = availabilityService
                .checkAvailabilityExcluding(availabilityRequest, id);

            if (!availability.getAvailable()) {
                throw new RuntimeException(
                    String.format("Pas assez de places disponibles. Places disponibles: %d, demandées: %d",
                        availability.getAvailableSeats(),
                        newNombrePersonnes)
                );
            }
        }

        // Mettre à jour les champs
        if (request.getEmail() != null) {
            reservation.setEmail(request.getEmail());
        }
        if (request.getTelephone() != null) {
            reservation.setTelephone(request.getTelephone());
        }
        if (request.getDateReservation() != null) {
            reservation.setDateReservation(request.getDateReservation());
        }
        if (request.getNombrePersonnes() != null) {
            reservation.setNombrePersonnes(request.getNombrePersonnes());
        }
        if (request.getCommentaire() != null) {
            reservation.setCommentaire(request.getCommentaire());
        }
        if (request.getPlatIds() != null && !request.getPlatIds().isEmpty()) {
            List<Plat> plats = platRepository.findAllById(request.getPlatIds());
            reservation.setPlats(plats);
        }

        Reservation updated = reservationRepository.save(reservation);
        return convertToResponse(updated);
    }

    /**
     * Récupère toutes les réservations en attente de validation.
     *
     * @return Liste des réservations avec statut EN_ATTENTE
     */
    public List<ReservationResponse> getPendingReservations() {
        return reservationRepository.findByStatut(StatutReservation.EN_ATTENTE).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Récupère les réservations par statut.
     *
     * @param status Statut à filtrer
     * @return Liste des réservations avec ce statut
     */
    public List<ReservationResponse> getReservationsByStatus(String status) {
        StatutReservation statutReservation = StatutReservation.valueOf(status.toUpperCase());
        return reservationRepository.findByStatut(statutReservation).stream()
                .map(this::convertToResponse)
                .collect(Collectors.toList());
    }

    /**
     * Approuve une réservation (change le statut à CONFIRMEE).
     *
     * @param id Identifiant de la réservation
     * @return Réservation approuvée
     * @throws RuntimeException si la réservation n'est pas trouvée
     */
    public ReservationResponse approveReservation(Long id) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Réservation non trouvée"));
        reservation.setStatut(StatutReservation.CONFIRMEE);
        return convertToResponse(reservationRepository.save(reservation));
    }

    /**
     * Rejette une réservation (change le statut à ANNULEE).
     *
     * @param id Identifiant de la réservation
     * @return Réservation rejetée
     * @throws RuntimeException si la réservation n'est pas trouvée
     */
    public ReservationResponse rejectReservation(Long id) {
        Reservation reservation = reservationRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Réservation non trouvée"));
        reservation.setStatut(StatutReservation.ANNULEE);
        return convertToResponse(reservationRepository.save(reservation));
    }

    /**
     * Convertit une entité Reservation en DTO pour l'exposition via l'API.
     * Transforme également la liste des plats en PlatDTO.
     *
     * @param reservation Entité réservation à convertir
     * @return DTO contenant toutes les informations de la réservation
     */
    private ReservationResponse convertToResponse(Reservation reservation) {
        // Conversion des plats en DTO
        List<PlatDTO> platDTOs = reservation.getPlats().stream()
                .map(plat -> PlatDTO.builder()
                        .id(plat.getId())
                        .nom(plat.getNom())
                        .description(plat.getDescription())
                        .prix(plat.getPrix())
                        .categorie(plat.getCategorie().name())
                        .build())
                .collect(Collectors.toList());

        // Construction de la réponse complète
        return ReservationResponse.builder()
                .id(reservation.getId())
                .idClient(reservation.getUser().getId())
                .email(reservation.getEmail())
                .telephone(reservation.getTelephone())
                .dateReservation(reservation.getDateReservation())
                .nombrePersonnes(reservation.getNombrePersonnes())
                .plats(platDTOs)
                .statut(reservation.getStatut().name())
                .commentaire(reservation.getCommentaire())
                .dateCreation(reservation.getDateCreation())
                .build();
    }
}

File: api_fastapi/app/services/reservation_service.py
from datetime import datetime
from sqlalchemy.orm import Session
from fastapi import HTTPException, status
from app.models.reservation import Reservation, StatutReservation
from app.models.plat import Plat
from app.schemas.reservation import (
    ReservationRequest,
    ReservationResponse,
    UpdateReservationRequest,
    AvailabilityRequest,
)
from app.services.availability_service import AvailabilityService


class ReservationService:
    """Service for reservation operations"""

    @staticmethod
    def create(db: Session, request: ReservationRequest) -> ReservationResponse:
        """Create a new reservation"""
        # Validate future date
        if request.dateReservation < datetime.now():
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="La date de réservation doit être dans le futur"
            )

        # Check availability
        availability_request = AvailabilityRequest(
            dateReservation=request.dateReservation,
            nombrePersonnes=request.nombrePersonnes
        )
        availability = AvailabilityService.check_availability(db, availability_request)

        if not availability.available:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Pas assez de places disponibles. Places restantes: {availability.availableSeats}"
            )

        # Get dishes if provided
        plats = []
        if request.platIds:
            plats = db.query(Plat).filter(Plat.id.in_(request.platIds)).all()
            if len(plats) != len(request.platIds):
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Un ou plusieurs plats n'existent pas"
                )

        # Create reservation
        reservation = Reservation(
            user_id=request.idClient,
            email=request.email,
            telephone=request.telephone,
            date_reservation=request.dateReservation,
            nombre_personnes=request.nombrePersonnes,
            statut=StatutReservation.EN_ATTENTE,
            commentaire=request.commentaire,
            date_creation=datetime.now(),
            plats=plats
        )

        db.add(reservation)
        db.commit()
        db.refresh(reservation)

        return ReservationResponse.from_orm(reservation)

    @staticmethod
    def get_all(db: Session) -> list[ReservationResponse]:
        """Get all reservations"""
        reservations = db.query(Reservation).all()
        return [ReservationResponse.from_orm(r) for r in reservations]

    @staticmethod
    def get_by_id(db: Session, reservation_id: int) -> ReservationResponse:
        """Get reservation by ID"""
        reservation = db.query(Reservation).filter(Reservation.id == reservation_id).first()
        if not reservation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Réservation avec id {reservation_id} non trouvée"
            )
        return ReservationResponse.from_orm(reservation)

    @staticmethod
    def get_by_email(db: Session, email: str) -> list[ReservationResponse]:
        """Get reservations by email"""
        reservations = db.query(Reservation).filter(Reservation.email == email).all()
        return [ReservationResponse.from_orm(r) for r in reservations]

    @staticmethod
    def get_by_user_id(db: Session, user_id: int) -> list[ReservationResponse]:
        """Get reservations by user ID"""
        reservations = db.query(Reservation).filter(Reservation.user_id == user_id).all()
        return [ReservationResponse.from_orm(r) for r in reservations]

    @staticmethod
    def get_by_status(db: Session, statut: StatutReservation) -> list[ReservationResponse]:
        """Get reservations by status"""
        reservations = db.query(Reservation).filter(Reservation.statut == statut).all()
        return [ReservationResponse.from_orm(r) for r in reservations]

    @staticmethod
    def get_pending(db: Session) -> list[ReservationResponse]:
        """Get all pending reservations"""
        return ReservationService.get_by_status(db, StatutReservation.EN_ATTENTE)

    @staticmethod
    def update(
        db: Session,
        reservation_id: int,
        request: UpdateReservationRequest
    ) -> ReservationResponse:
        """Update an existing reservation"""
        reservation = db.query(Reservation).filter(Reservation.id == reservation_id).first()
        if not reservation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Réservation avec id {reservation_id} non trouvée"
            )

        # Check availability if date or number of people is being changed
        if request.dateReservation or request.nombrePersonnes:
            new_date = request.dateReservation or reservation.date_reservation
            new_people = request.nombrePersonnes or reservation.nombre_personnes

            availability_request = AvailabilityRequest(
                dateReservation=new_date,
                nombrePersonnes=new_people
            )
            availability = AvailabilityService.check_availability(
                db, availability_request, exclude_reservation_id=reservation_id
            )

            if not availability.available:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Pas assez de places disponibles. Places restantes: {availability.availableSeats}"
                )

        # Update fields
        if request.email:
            reservation.email = request.email
        if request.telephone:
            reservation.telephone = request.telephone
        if request.dateReservation:
            reservation.date_reservation = request.dateReservation
        if request.nombrePersonnes:
            reservation.nombre_personnes = request.nombrePersonnes
        if request.commentaire is not None:
            reservation.commentaire = request.commentaire

        # Update dishes if provided
        if request.platIds is not None:
            plats = db.query(Plat).filter(Plat.id.in_(request.platIds)).all()
            if len(plats) != len(request.platIds):
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Un ou plusieurs plats n'existent pas"
                )
            reservation.plats = plats

        db.commit()
        db.refresh(reservation)

        return ReservationResponse.from_orm(reservation)

    @staticmethod
    def update_status(
        db: Session,
        reservation_id: int,
        statut: StatutReservation
    ) -> ReservationResponse:
        """Update reservation status"""
        reservation = db.query(Reservation).filter(Reservation.id == reservation_id).first()
        if not reservation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Réservation avec id {reservation_id} non trouvée"
            )

        reservation.statut = statut
        db.commit()
        db.refresh(reservation)

        return ReservationResponse.from_orm(reservation)

    @staticmethod
    def cancel(db: Session, reservation_id: int) -> None:
        """Cancel a reservation (soft delete)"""
        reservation = db.query(Reservation).filter(Reservation.id == reservation_id).first()
        if not reservation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Réservation avec id {reservation_id} non trouvée"
            )

        reservation.statut = StatutReservation.ANNULEE
        db.commit()

    @staticmethod
    def approve(db: Session, reservation_id: int) -> ReservationResponse:
        """Approve a pending reservation"""
        return ReservationService.update_status(db, reservation_id, StatutReservation.CONFIRMEE)

    @staticmethod
    def reject(db: Session, reservation_id: int) -> ReservationResponse:
        """Reject a pending reservation"""
        return ReservationService.update_status(db, reservation_id, StatutReservation.ANNULEE)
